File: api/stocks.py
from http.server import BaseHTTPRequestHandler
import json
import sys
import os
from urllib.parse import urlparse, parse_qs

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

from app.services.stock_data import get_all_stocks, get_stock_history
from app.services.analysis import score_stock


class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

    def do_GET(self):
        try:
            parsed = urlparse(self.path)
            params = parse_qs(parsed.query)
            sort = params.get('sort', ['score'])[0]
            order = params.get('order', ['desc'])[0]
            limit = int(params.get('limit', ['100'])[0])
            keyword = params.get('keyword', [''])[0]

            df = get_all_stocks()

            if keyword:
                mask = df['code'].str.contains(keyword, na=False) | df['name'].str.contains(keyword, na=False)
                df = df[mask]

            results = []
            for _, row in df.iterrows():
                code = str(row['code'])
                try:
                    hist = get_stock_history(code, days=60)
                    scoring = score_stock(hist)
                except Exception:
                    scoring = {"score": 50, "details": {}, "signals": []}

                results.append({
                    "code": code,
                    "name": row['name'],
                    "price": float(row.get('price', 0)),
                    "change_pct": float(row.get('change_pct', 0)),
                    "change_amount": float(row.get('change_amount', 0)),
                    "volume": float(row.get('volume', 0)),
                    "turnover": float(row.get('turnover', 0)),
                    "pe": None, "pb": None, "turnover_rate": 0,
                    "score": scoring["score"],
                    "signals": scoring["signals"],
                })

            reverse = order == "desc"
            if sort == "score":
                results.sort(key=lambda x: x["score"], reverse=reverse)
            elif sort == "change_pct":
                results.sort(key=lambda x: x["change_pct"], reverse=reverse)
            elif sort == "volume":
                results.sort(key=lambda x: x["volume"], reverse=reverse)

            self._respond(200, {"ok": True, "data": results[:limit], "total": len(results)})
        except Exception as e:
            self._respond(500, {"ok": False, "detail": str(e)})

    def _respond(self, code, data):
        self.send_response(code)
        self.send_header('Content-type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        self.wfile.write(json.dumps(data, ensure_ascii=False).encode())

File: api/stocks.js
import { fetchAllStocks, fetchStockHistory } from '../../lib/stock-data.js';
import { scoreStock, quickScore } from '../../lib/analysis.js';

export default async function handler(req, res) {
  try {
    const { sort = 'score', order = 'desc', limit = 100, keyword = '' } = req.query;
    let stocks = await fetchAllStocks();

    if (keyword) {
      const kw = keyword.toLowerCase();
      stocks = stocks.filter(s => s.code.includes(kw) || s.name.toLowerCase().includes(kw));
    }

    const results = await Promise.all(stocks.slice(0, +limit * 2).map(async (s) => {
      let scoring;
      try {
        const hist = await fetchStockHistory(s.code, 60);
        scoring = scoreStock(hist);
      } catch {
        scoring = quickScore(s);
      }
      return {
        code: s.code, name: s.name, price: s.price,
        change_pct: s.change_pct, change_amount: s.change_amount,
        volume: s.volume, turnover: s.turnover || 0,
        pe: null, pb: null, turnover_rate: 0,
        score: scoring.score, signals: scoring.signals,
      };
    }));

    const reverse = order === 'desc';
    if (sort === 'score') results.sort((a, b) => reverse ? b.score - a.score : a.score - b.score);
    else if (sort === 'change_pct') results.sort((a, b) => reverse ? b.change_pct - a.change_pct : a.change_pct - b.change_pct);
    else if (sort === 'volume') results.sort((a, b) => reverse ? b.volume - a.volume : a.volume - b.volume);

    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Cache-Control', 's-maxage=60');
    res.status(200).json({ ok: true, data: results.slice(0, +limit), total: results.length });
  } catch (e) {
    res.status(500).json({ ok: false, detail: e.message });
  }
}
